servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java

```diff
@@ -359,12 +483,32 @@ void markInactive() {
             }
         }
 
+        @SuppressWarnings("PMD.CollapsibleIfStatements")
+        void markExpired() {
+            stateUpdater.set(this, connections.length == 0 ? State.CLOSED : State.EXPIRED);
```

Comment:
since there are two independent volatiles involved, can you add a comment to clarify why accessing independent volatile state is safe?

Reply:
Please review the added comment.

Reply:
I think we have a race condition which may lea
```
Host A {connection1, connection2}

T1 - SD event {A unavailable}, A.markExpired(), connections.length == 2
T2 - c1 closed, existing.length == 2, state == ACTIVE, new = {c2}, connectionsUpdater.cas(old, new)
T2 - c2 closed, existing.length == 1, state == ACTIVE, new = {}, connectionsUpdater.cas(old, new)
T1 - state = expired
```

note that `Host#closeAsync()` is never called, and `createHost`'s `host.onClose().afterFinally(() -> ..)` may never run to cleanup `usedHostsUpdater`. I think there are also scenarios where we may create connections after SD has told us the host is no longer available. I think we should stick to a single variable for the state and connections, otherwise it is difficult to eliminate race conditions.

```java
        @SuppressWarnings("rawtypes")
        private static final AtomicReferenceFieldUpdater<Host, ConnState> connStateUpdater =
                AtomicReferenceFieldUpdater.newUpdater(Host.class, ConnState.class, "connState");
        private static final class ConnState {
            final Object[] connections;
            final State state;

            ConnectionsState(Object[] connections, State state) {
                this.connections = connections;
                this.state = state;
            }
        }
        private static final ConnState EMPTY_CONN_STATE = new ConnState(EMPTY_ARRAY, ...);
        
        private volatile ConnState connState = EMPTY_CONN_STATE;
```

Reply:
Thank you for spotting that and for the excellent suggestion. I've implemented it the way you suggest, will push a commit soon.